README.md
# Scala Binding for ZeroMQ

The Scala binding for ZeroMQ is based on ZeroMQ versions 3.2.0 and uses [JNA][]
for accessing native functions. It is a thin wrapper of the ZeroMQ API, but
offers also a `jzmq`-like API intending to be compatible with the Java binding
for ZeroMQ; users of the Java binding wanting to access ZeroMQ over JNA can
switch over to the Scala binding.

[JNA]: https://github.com/twall/jna

## Installation

Scala binding for ZeroMQ is made available through a Maven repository. If
you are using SBT, amend your `build.sbt` or Build.scala with:

````
resolvers += "Sonatype (releases)" at "https://oss.sonatype.org/content/repositories/releases/"

libraryDependencies += "org.zeromq" %% "zeromq-scala-binding" % "1.0.0-SNAPSHOT"
````


Changes.md
# Changes

## 3.2 Upgrade API Changes
See http://www.zeromq.org/docs:3-1-upgrade for documented ZeroMQ modifications, removals, additions.

The following changes were tested against ZeroMQ 2.2.0 adn 3.2.0 on MacOSX 10.7.5. Further
testing on a other operating systems is needed.

### Modifications
Product version changed to 1.0.0-SNAPSHOT as an attempt to not collide with < 3.2.0 binding upgrades.

The scala version has been upgraded to 2.10.0-M7

The sbt version has been upgraded to 0.12.0

Six method return types have been modified to expose the zmq lib `int` success/failure, which
were not exposed in previous versions:

org.zeromq.Socket.java [migrated from `void` => `int`]:
````
int destroy()

int bind(String addr)

int connect(String addr)

int subscribe(byte[] topic)

int unsubscribe(byte[] topic)

int setBytesSockopt(int option, byte[] optval)

````

Copyright updated to 2012

### Additions
Added sbtscalariform plugin

org.zeromq.Context.scala
````
int destroy()
````

org.zeromq.ZeroMQLibrary.scala
````
def zmq_ctx_new: Pointer // creates with default thread count of 1
def zmq_ctx_set(socket: Pointer, name: Int, value: Int): Int
def zmq_msg_send(msg: zmq_msg_t, socket: Pointer, flags: Int): Int
def zmq_msg_recv(msg: zmq_msg_t, socket: Pointer, flags: Int): Int
def zmq_ctx_destroy(context: Pointer): Int

````

### Removals
org.zeromq.Context.java: unused & unimplemented

````
void term()
````

src/test/scala/org/zeromq/ZMQSpec.scala
/*
 * Copyright 2011 - 2012 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.zeromq

import org.zeromq.ZMQ._
import org.zeromq.ZeroMQ._

class ZMQSpec extends AbstractZeroMQSpec {
  "ZMQ" must {
    "be successfully created and destroyed" in {
      val context = new Context(1)
      context must not be (null)
      context.poller(1) must not be (null)
      val s = context.socket(PULL)
      s must not be (null)
      s.close
      context.destroy() must be(0)
    }
    "support Socket#getType" in {
      val context = new Context()
      val sub = context.socket(SUB)
      sub.getType must equal(SUB)
      sub.close
      context.destroy()
    }
    "support pub-sub connection pattern" in {
      val context = new Context()
      val (pub, sub, poller) = (
        context.socket(PUB),
        context.socket(SUB),
        context.poller)
      pub.bind(endpoint) must be(0)
      sub.connect(endpoint) must be(0)
      sub.subscribe(subscribeAll) must be(0)
      poller.register(sub) must be(0)
      pub.send(dataBytes, ZMQ_DONTWAIT)
      poller.poll must equal(1)
      poller.pollin(0) must equal(true)
      sub.recv(0) must equal(dataBytes)
      sub.close
      pub.close
      context.destroy()
    }
    "support polling of multiple sockets" in {
      val context = new Context()
      val (pub, poller) = (context.socket(PUB), context.poller)
      pub.bind(endpoint)
      val (sub_x, sub_y) = (connectTestSubscriber(context), connectTestSubscriber(context))
      poller.register(sub_x)
      poller.register(sub_y)
      pub.send(dataBytes, 0)
      poller.poll must equal(2)
      poller.pollin(0) must equal(true)
      poller.pollin(1) must equal(true)
      sub_x.close
      sub_y.close
      pub.close
      context.destroy()
    }
    "support sending of zero-length messages" in {
      val context = new Context()
      val pub = context.socket(PUB)
      pub.bind(endpoint)
      pub.send("".getBytes, 0) must be(true)
      pub.close
      context.destroy()
    }
    "support socket linger" in {
      val context = new Context()
      val socket = context.socket(PUB)
      socket.setLinger(1000)
      // FIXME fails: socket.getLinger must be(0)
      socket.close
      context.destroy()
    }
    "support socket backlog" in {
      val context = new Context()
      val socket = context.socket(REP)
      socket.setBacklog(200)
      socket.getBacklog must be(100)
      socket.close
      socket.connect("inproc://reqrep")
      socket.getBacklog must be(0)
      socket.close
      context.destroy()
    }
    "support socket reconnect interval" in {
      val context = new Context()
      val socket = context.socket(REP)
      socket.setReconnectIVL(101)
      socket.getReconnectIVL must be(100)
      socket.close
      socket.connect("inproc://reqrep")
      socket.getReconnectIVL must be(0)
      socket.close
      context.destroy()
    }
    /* TODO test coverage:
    getReconnectIVLMax
    getMaxMsgSize
    getSndHWM
    getRcvHWM
    getHWM
    getSwap
    getAffinity
    getIdentity
    getRate
    getRecoveryInterval
    getMulticastHops
    getReceiveTimeOut
    getSendTimeOut
    getSendBufferSize
    getReceiveBufferSize
    getFD
    getEvents

    hasMulticastLoop
    hasReceiveMore

    setReceiveTimeOut
    setMulticastHops
    setReceiveTimeOut
    setSendTimeOut  
    setReconnectIVL  
    setReconnectIVLMax 
    setMaxMsgSize 
    setSndHWM 
    setRcvHWM  
    setHWM(long hwm)  
    setSwap 
    setAffinity
    setIdentity
    setRate 
    setRecoveryInterval 
    setMulticastLoop 
    setSendBufferSize 
    setReceiveBufferSize
    */
  }
}


src/main/scala/org/zeromq/ZeroMQ.scala
/*
 * Copyright 2011 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.zeromq

import com.sun.jna._
import com.sun.jna.ptr._

object ZeroMQ {
  /** Socket types */
  val ZMQ_PAIR = 0
  val ZMQ_PUB = 1
  val ZMQ_SUB = 2
  val ZMQ_REQ = 3
  val ZMQ_REP = 4
  val ZMQ_DEALER = 5
  val ZMQ_ROUTER = 6
  val ZMQ_PULL = 7
  val ZMQ_PUSH = 8
  val ZMQ_XPUB = 9
  val ZMQ_XSUB = 10
  /** Socket options */
  val ZMQ_HWM = 1
  val ZMQ_SWAP = 3
  val ZMQ_AFFINITY = 4
  val ZMQ_IDENTITY = 5
  val ZMQ_SUBSCRIBE = 6
  val ZMQ_UNSUBSCRIBE = 7
  val ZMQ_RATE = 8
  val ZMQ_RECOVERY_IVL = 9
  val ZMQ_MCAST_LOOP = 10
  val ZMQ_SNDBUF = 11
  val ZMQ_RCVBUF = 12
  val ZMQ_RCVMORE = 13
  val ZMQ_FD = 14
  val ZMQ_EVENTS = 15
  val ZMQ_TYPE = 16
  val ZMQ_LINGER = 17
  val ZMQ_RECONNECT_IVL = 18
  val ZMQ_BACKLOG = 19
  val ZMQ_RECONNECT_IVL_MAX = 21
  /** Built-in devices */
  val ZMQ_STREAMER = 1
  val ZMQ_FORWARDER = 2
  val ZMQ_QUEUE = 3
  /** Unix errors */
  val EINVAL = 22
  /** ZMQ errors */
  val ZMQ_HAUSNUMERO = 156384712
  val EFSM = ZMQ_HAUSNUMERO + 51
  val ENOCOMPATPROTO = ZMQ_HAUSNUMERO + 52
  val ETERM = ZMQ_HAUSNUMERO + 53
  /** ZMQ message definition */
  val ZMQ_MAX_VSM_SIZE = 30
  val ZMQ_DELIMITER = 31
  val ZMQ_VSM = 32
  val ZMQ_MSG_MORE = 1
  val ZMQ_MSG_SHARED = 128
  val ZMQ_MSG_MASK = 129
  /** IO multiplexing */
  val ZMQ_POLLIN: Short = 1
  val ZMQ_POLLOUT: Short = 2
  val ZMQ_POLLERR: Short = 4 
  /** Helper for loading the ZeroMQ library */
  def loadLibrary: ZeroMQ = {
    Native.loadLibrary("zmq", classOf[ZeroMQ]).asInstanceOf[ZeroMQ]  
  }
}

trait ZeroMQ extends Library {
  def zmq_bind(socket: Pointer, endpoint: String): Int
  def zmq_close(socket: Pointer): Int
  def zmq_connect(scoket: Pointer, endpoint: String): Int
  def zmq_device(device: Int, frontend: Pointer, backend: Pointer): Int
  def zmq_errno: Int
  def zmq_getsockopt(socket: Pointer, option_name: Int, option_value: Pointer, option_len: LongByReference): Int
  def zmq_init(io_threads: Int): Pointer
  def zmq_msg_init(msg: zmq_msg_t): Int
  def zmq_msg_close(msg: zmq_msg_t): Int
  def zmq_msg_copy(dest: zmq_msg_t, src: zmq_msg_t): Int
  def zmq_msg_data(msg: zmq_msg_t): Pointer
  def zmq_msg_init_data(msg: zmq_msg_t, data: Pointer, size: NativeLong, ffn: zmq_free_fn, hint: Pointer): Int
  def zmq_msg_init_size(msg: zmq_msg_t, size: NativeLong): Int
  def zmq_msg_move(dest: zmq_msg_t, src: zmq_msg_t): Int
  def zmq_poll(items: Array[zmq_pollitem_t], nitems: Int, timeout: NativeLong): Int
  def zmq_recv(socket: Pointer, msg: zmq_msg_t, flags: Int): Int
  def zmq_send(socket: Pointer, msg: zmq_msg_t, flags: Int): Int
  def zmq_setsockopt(socket: Pointer, option_name: Int, option_value: Pointer, option_len: NativeLong): Int
  def zmq_socket(context: Pointer, socket_type: Int): Pointer
  def zmq_strerror(errnum: Int): String
  def zmq_term(context: Pointer): Int
  def zmq_version(major: Array[Int], minor: Array[Int], patch: Array[Int]): Unit
}


src/main/java/org/zeromq/ZMQ.java
/*
 * Copyright 2011 - 2012 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.zeromq;

import com.sun.jna.Memory;
import com.sun.jna.NativeLong;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.LongByReference;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;

/**
 * Offers an API similar to that of jzmq [1] written by Gonzalo Diethelm.
 * <p/>
 * 1. https://github.com/zeromq/jzmq
 */
public class ZMQ {
  private static final ZeroMQLibrary zmq = ZeroMQ$.MODULE$.loadLibrary();
  private static final int[] majorVersion = new int[1];
  private static final int[] minorVersion = new int[1];
  private static final int[] patchVersion = new int[1];

  public static final int NOBLOCK   = ZeroMQ$.MODULE$.ZMQ_NOBLOCK();
  public static final int DONTWAIT  = ZeroMQ$.MODULE$.ZMQ_DONTWAIT();
  public static final int PAIR      = ZeroMQ$.MODULE$.ZMQ_PAIR();
  public static final int SNDMORE   = ZeroMQ$.MODULE$.ZMQ_SNDMORE();
  public static final int PUB       = ZeroMQ$.MODULE$.ZMQ_PUB();
  public static final int SUB       = ZeroMQ$.MODULE$.ZMQ_SUB();
  public static final int REQ       = ZeroMQ$.MODULE$.ZMQ_REQ();
  public static final int REP       = ZeroMQ$.MODULE$.ZMQ_REP();
  public static final int XREQ      = ZeroMQ$.MODULE$.ZMQ_DEALER();
  public static final int XREP      = ZeroMQ$.MODULE$.ZMQ_ROUTER();
  public static final int DEALER    = ZeroMQ$.MODULE$.ZMQ_DEALER();
  public static final int ROUTER    = ZeroMQ$.MODULE$.ZMQ_ROUTER();
  public static final int PULL      = ZeroMQ$.MODULE$.ZMQ_PULL();
  public static final int PUSH      = ZeroMQ$.MODULE$.ZMQ_PUSH();
  public static final int STREAMER  = ZeroMQ$.MODULE$.ZMQ_STREAMER();
  public static final int FORWARDER = ZeroMQ$.MODULE$.ZMQ_FORWARDER();
  public static final int QUEUE     = ZeroMQ$.MODULE$.ZMQ_QUEUE();

  static {
    zmq.zmq_version(majorVersion, minorVersion, patchVersion);
  }

  public static byte[] intToByteArray(int value) {
    return new byte[]{(byte) (value >>> 24), (byte) (value >>> 16), (byte) (value >>> 8), (byte) value};
  }

  public static int getMajorVersion() {
    return majorVersion[0];
  }

  public static int getMinorVersion() {
    return minorVersion[0];
  }

  public static int getPatchVersion() {
    return patchVersion[0];
  }

  public static int getFullVersion() {
    return makeVersion(getMajorVersion(), getMinorVersion(), getPatchVersion());
  }

  public static int makeVersion(int major, int minor, int patch) {
    return major * 10000 + minor * 100 + patch;
  }

  public static String getVersionString() {
    return String.format("%d.%d.%d", getMajorVersion(), getMinorVersion(), getPatchVersion());
  }

  public static Context context(int ioThreads) {
    return new Context(ioThreads);
  }

  /**
   * Creates a new Context with the default thread count of 1.
   */
  public static Context context() {
    return new Context();
  }

  /**
   * Represents a ZeroMQ Context
   */
  public static class Context {
    protected Pointer ptr;


    protected Context(int ioThreads) {
      ptr = (getFullVersion() < makeVersion(3, 2, 0)) ? zmq.zmq_init(ioThreads) : zmq.zmq_ctx_new();
    }

    /**
     * Creates a new Context with the default number of threads.
     */
    protected Context() {
      ptr = (getFullVersion() < makeVersion(3, 2, 0)) ? zmq.zmq_init(1) : zmq.zmq_ctx_new();
    }

    public Socket socket(int type) {
      return new Socket(this, type);
    }

    public Poller poller() {
      return new Poller(this);
    }

    public Poller poller(int size) {
      return new Poller(this, size);
    }

    /**
     * Returns zero if successful, otherwise returns -1.
     */
    public int destroy() {
      return (getFullVersion() < makeVersion(3, 2, 0)) ? zmq.zmq_term(ptr) : zmq.zmq_ctx_destroy(ptr);
    }
  }

  public static class Socket {
    protected Pointer ptr;
    MessageDataBuffer messageDataBuffer = new MessageDataBuffer();

    public void close() {
      zmq.zmq_close(ptr);
    }

    public int getType() {
      return (getFullVersion() < makeVersion(2, 1, 0)) ? -1 : (int) getLongSockopt(ZeroMQ$.MODULE$.ZMQ_TYPE());
    }

    public void setLinger(long linger) {
      if (getFullVersion() >= makeVersion(2, 1, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_LINGER(), linger);
    }

    public long getLinger() {
      return (getFullVersion() >= makeVersion(2, 1, 0)) ? (int) getLongSockopt(ZeroMQ$.MODULE$.ZMQ_LINGER()) : -1;
    }

    public void setBacklog(long backlog) {
      if (getFullVersion() >= makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_BACKLOG(), backlog);
    }

    public long getBacklog() {
      return (getFullVersion() >= makeVersion(3, 0, 0)) ? (int) getLongSockopt(ZeroMQ$.MODULE$.ZMQ_BACKLOG()) : -1;
    }

    public void setReconnectIVL(long reconnectIVL) {
      if (getFullVersion() >= makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_RECONNECT_IVL(), reconnectIVL);
    }

    public long getReconnectIVL() {
      return (getFullVersion() >= makeVersion(3, 0, 0)) ? (int) getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RECONNECT_IVL()) : -1;
    }

    public void setReconnectIVLMax(long reconnectIVLMax) {
      if (getFullVersion() >= makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_RECONNECT_IVL_MAX(), reconnectIVLMax);
    }

    public long getReconnectIVLMax() {
      return (getFullVersion() >= makeVersion(3, 0, 0)) ? getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RECONNECT_IVL_MAX()) : -1;
    }

    public void setMaxMsgSize(long maxMsgSize) {
      if (getFullVersion() >= makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_MAXMSGSIZE(), maxMsgSize);
    }

    public long getMaxMsgSize() {
      return (getFullVersion() >= makeVersion(3, 0, 0)) ? getLongSockopt(ZeroMQ$.MODULE$.ZMQ_MAXMSGSIZE()) : -1;
    }

    public void setHWM(long hwm) {
      if (getFullVersion() < makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_HWM(), hwm);
      else {
        setSndHWM(hwm);
        setRcvHWM(hwm);
      }
    }

    public long getHWM() {
      return (getFullVersion() < makeVersion(3, 0, 0)) ? getLongSockopt(ZeroMQ$.MODULE$.ZMQ_HWM()) : -1;
    }

    public void setSndHWM(long sndHWM) {
      if (getFullVersion() >= makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_SNDHWM(), sndHWM);
    }

    public long getSndHWM() {
      return (getFullVersion() >= makeVersion(3, 0, 0)) ? getLongSockopt(ZeroMQ$.MODULE$.ZMQ_SNDHWM()) : -1;
    }

    public void setRcvHWM(long rcvHWM) {
      if (getFullVersion() >= makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_RCVHWM(), rcvHWM);
    }

    public long getRcvHWM() {
      return (getFullVersion() >= makeVersion(3, 0, 0)) ? getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RCVHWM()) : -1;
    }


    public void setSwap(long swap) {
      if (getFullVersion() < makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_SWAP(), swap);
    }

    public long getSwap() {
      return (getFullVersion() < makeVersion(3, 0, 0)) ? getLongSockopt(ZeroMQ$.MODULE$.ZMQ_SWAP()) : -1;
    }

    public void setAffinity(long affinity) {
      setLongSockopt(ZeroMQ$.MODULE$.ZMQ_AFFINITY(), affinity);
    }

    public long getAffinity() {
      return getLongSockopt(ZeroMQ$.MODULE$.ZMQ_AFFINITY());
    }

    public void setIdentity(byte[] identity) {
      setBytesSockopt(ZeroMQ$.MODULE$.ZMQ_IDENTITY(), identity);
    }

    public byte[] getIdentity() {
      return getBytesSockopt(ZeroMQ$.MODULE$.ZMQ_IDENTITY());
    }

    public void setRate (long rate) {
      setLongSockopt(ZeroMQ$.MODULE$.ZMQ_RATE(), rate);
    }

    public long getRate() {
      return getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RATE());
    }

    public void setRecoveryInterval(long recovery_ivl) {
      setLongSockopt(ZeroMQ$.MODULE$.ZMQ_RECONNECT_IVL(), recovery_ivl);
    }

    public long getRecoveryInterval() {
      return getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RECOVERY_IVL());
    }

    public void setMulticastHops(long mcast_hops) {
      if (getFullVersion() >= makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_MCAST_LOOP(), mcast_hops);
    }

    public long getMulticastHops() {
      return (getFullVersion() < makeVersion(3, 0, 0)) ? -1 : getLongSockopt(ZeroMQ$.MODULE$.ZMQ_MCAST_LOOP());
    }

    public void setReceiveTimeOut(long timeout) {
      if (getFullVersion() >= makeVersion(2, 2, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_RCVTIMEO(), timeout);
    }

    public long getReceiveTimeOut() {
      return (getFullVersion() < makeVersion(2, 2, 0)) ? -1 : getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RCVTIMEO());
    }

    public void setSendTimeOut(long timeout) {
      if (getFullVersion() >= makeVersion(2, 2, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_SNDTIMEO(), timeout);
    }

    public long getSendTimeOut() {
      return (getFullVersion() < makeVersion(2, 2, 0)) ? -1 : getLongSockopt(ZeroMQ$.MODULE$.ZMQ_SNDTIMEO());
    }

    public void setSendBufferSize(long sndbuf) {
      setLongSockopt(ZeroMQ$.MODULE$.ZMQ_SNDBUF(), sndbuf);
    }

    public long getSendBufferSize() {
      return getLongSockopt(ZeroMQ$.MODULE$.ZMQ_SNDBUF());
    }

    public void setReceiveBufferSize(long rcvbuf) {
      setLongSockopt(ZeroMQ$.MODULE$.ZMQ_RCVBUF(), rcvbuf);
    }

    public long getReceiveBufferSize() {
      return getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RCVBUF());
    }

    public long getFD() {
      return (getFullVersion() < makeVersion(2, 1, 0)) ? -1 : getLongSockopt(ZeroMQ$.MODULE$.ZMQ_FD());
    }

    public long getEvents() {
      return (getFullVersion() < makeVersion(2, 1, 0)) ? -1 : getLongSockopt(ZeroMQ$.MODULE$.ZMQ_EVENTS());
    }

    public void setMulticastLoop(boolean mcast_loop) {
      if (getFullVersion() < makeVersion(3, 0, 0)) setLongSockopt(ZeroMQ$.MODULE$.ZMQ_MCAST_LOOP(), mcast_loop ? 1 : 0);
    }

    public boolean hasMulticastLoop() {
      return getFullVersion() >= makeVersion(3, 0, 0) && getLongSockopt(ZeroMQ$.MODULE$.ZMQ_MCAST_LOOP()) != 0;
    }

    public boolean hasReceiveMore() {
      return getLongSockopt(ZeroMQ$.MODULE$.ZMQ_RCVMORE()) != 0;
    }

    public int subscribe(byte[] topic) {
      return setBytesSockopt(ZeroMQ$.MODULE$.ZMQ_SUBSCRIBE(), topic);
    }

    public int unsubscribe(byte[] topic) {
      return setBytesSockopt(ZeroMQ$.MODULE$.ZMQ_UNSUBSCRIBE(), topic);
    }

    public int bind(String addr) {
        return zmq.zmq_bind(ptr, addr);
    }

    public int connect(String addr) {
        return zmq.zmq_connect(ptr, addr);
    }

    public boolean send(byte[] msg, int flags) {
      zmq_msg_t message = newZmqMessage(msg);

      if ((getFullVersion() == makeVersion(3, 2, 0) && zmq.zmq_msg_send(message, ptr, flags) == -1) ||
          (getFullVersion() < makeVersion(3, 2, 0) && zmq.zmq_send(ptr, message, flags) != 0)) {
        if (zmq.zmq_errno() == ZeroMQ$.MODULE$.EAGAIN()) {
          if (zmq.zmq_msg_close(message) != 0) raiseZMQException();
          else return false;
        } else {
          zmq.zmq_msg_close(message);
          raiseZMQException();
          return false;
        }
      }

      if (zmq.zmq_msg_close(message) != 0) raiseZMQException();
      return true;
    }

    public byte[] recv(int flags) {
      zmq_msg_t message = newZmqMessage();

      if ((getFullVersion() == makeVersion(3, 2, 0) && zmq.zmq_msg_recv(message, ptr, flags) == -1) ||
          (getFullVersion() < makeVersion(3, 2, 0) && zmq.zmq_recv(ptr, message, flags) != 0)) {
        if (zmq.zmq_errno() == ZeroMQ$.MODULE$.EAGAIN()) {
          if (zmq.zmq_msg_close(message) != 0) raiseZMQException();
          else return null;
        } else {
          zmq.zmq_msg_close(message);
          raiseZMQException();
        }
      }

      Pointer data = zmq.zmq_msg_data(message);
      int length = zmq.zmq_msg_size(message);
      byte[] dataByteArray = data.getByteArray(0, length);

      if (zmq.zmq_msg_close(message) != 0) raiseZMQException();
      return dataByteArray;
    }

    protected Socket(Context context, int type) {
      ptr = zmq.zmq_socket(context.ptr, type);
    }

    @Override
    protected void finalize() {
      close();
    }

    private long getLongSockopt(int option) {
      Memory value = new Memory(Long.SIZE / 8);
      zmq.zmq_getsockopt(ptr, option, value, new LongByReference(Long.SIZE / 8));
      return value.getLong(0);
    }

    private int setLongSockopt(int option, long optval) {
      Memory value = new Memory(Long.SIZE / 8);
      value.setLong(0, optval);
      return zmq.zmq_setsockopt(ptr, option, value, new NativeLong(Long.SIZE / 8));
    }

    private byte[] getBytesSockopt(int option) {
      Memory value = new Memory(1024);
      LongByReference length = new LongByReference(1024);
      zmq.zmq_getsockopt(ptr, option, value, length);
      return value.getByteArray(0, (int) length.getValue());
    }

    private int setBytesSockopt(int option, byte[] optval) {
      Pointer value;
      if (optval.length > 0) {
        value = new Memory(optval.length);
        value.write(0, optval, 0, optval.length);
      } else value = Pointer.NULL;

      return zmq.zmq_setsockopt(ptr, option, value, new NativeLong(optval.length));
    }

    private zmq_msg_t newZmqMessage(byte[] msg) {
      zmq_msg_t message = new zmq_msg_t();
      if (msg.length == 0) {
        if (zmq.zmq_msg_init_size(message, new NativeLong(msg.length)) != 0) raiseZMQException();
      } else {
        Memory mem = new Memory(msg.length);
        mem.write(0, msg, 0, msg.length);
        if (zmq.zmq_msg_init_data(message, mem, new NativeLong(msg.length), messageDataBuffer, mem) == 0)
          messageDataBuffer.add(mem);
        else raiseZMQException();
      }
      return message;
    }

    private zmq_msg_t newZmqMessage() {
      zmq_msg_t message = new zmq_msg_t();
      if (zmq.zmq_msg_init(message) != 0) raiseZMQException();
      return message;
    }

    private void raiseZMQException() {
      int errno = zmq.zmq_errno();
      throw new ZMQException(zmq.zmq_strerror(errno), errno);
    }

    private class MessageDataBuffer implements zmq_free_fn {
      private HashSet<Pointer> buffer = new HashSet<Pointer>();

      public synchronized void add(Pointer data) {
        buffer.add(data);
      }

      public synchronized void invoke(Pointer data, Pointer memory) {
        buffer.remove(memory);
      }
    }
  }

  public static class Poller {
    public static final int POLLIN = ZeroMQ$.MODULE$.ZMQ_POLLIN();
    public static final int POLLOUT = ZeroMQ$.MODULE$.ZMQ_POLLOUT();
    public static final int POLLERR = ZeroMQ$.MODULE$.ZMQ_POLLERR();

    private static final int SIZE_DEFAULT = 32;
    private static final int SIZE_INCREMENT = 16;
    private static final int UNINITIALIZED_TIMEOUT = -2;

    private long timeout = UNINITIALIZED_TIMEOUT;
    private int nextEventIndex = 0;
    private int maxEventCount = 0;
    private int curEventCount = 0;
    private Socket[] sockets = null;
    private short[] events = null;
    private short[] revents = null;
    private LinkedList<Integer> freeSlots = null;

    public int register(Socket socket) {
      return register(socket, POLLIN | POLLOUT | POLLERR);
    }

    public int register(Socket socket, int numEvents) {
      int pos = -1;
      if (!freeSlots.isEmpty()) pos = freeSlots.remove();
      else {
        if (nextEventIndex >= maxEventCount) {
          int newMaxEventCount = maxEventCount + SIZE_INCREMENT;
          sockets = Arrays.copyOf(sockets, newMaxEventCount);
          events = Arrays.copyOf(events, newMaxEventCount);
          revents = Arrays.copyOf(revents, newMaxEventCount);
          maxEventCount = newMaxEventCount;
        }
        pos = nextEventIndex++;
      }
      sockets[pos] = socket;
      events[pos] = (short) numEvents;
      curEventCount++;
      return pos;
    }

    public void unregister(Socket socket) {
      for (int index = 0; index < nextEventIndex; index++)
        if (sockets[index] == socket) {
          unregisterSocketAtIndex(index);
          break;
        }
    }

    private void unregisterSocketAtIndex(int index) {
      sockets[index] = null;
      events[index] = 0;
      revents[index] = 0;
      freeSlots.add(index);
      curEventCount--;
    }

    public Socket getSocket(int index) {
      return (index < 0 || index >= nextEventIndex) ? null : sockets[index];
    }

    public long getTimeout() {
      return timeout;
    }

    public void setTimeout(long timeout) {
      this.timeout = timeout;
    }

    public int getSize() {
      return maxEventCount;
    }

    public int getNext() {
      return nextEventIndex;
    }

    public long poll() {
      long timeout = -1;
      if (this.timeout != UNINITIALIZED_TIMEOUT) timeout = this.timeout;
      return poll(timeout);
    }

    public long poll(long timeout) {
      int pollItemCount = 0;
      for (int i = 0; i < nextEventIndex; i++) revents[i] = 0;

      if (curEventCount == 0) return 0;

      zmq_pollitem_t[] items = (zmq_pollitem_t[]) new zmq_pollitem_t().toArray(curEventCount);

      for (int i = 0; i < pollItemCount; i++) items[i] = new zmq_pollitem_t();

      for (int socketIndex = 0; socketIndex < sockets.length; socketIndex++) {
        if (sockets[socketIndex] == null) continue;

        items[pollItemCount].socket = sockets[socketIndex].ptr;
        items[pollItemCount].fd = 0;
        items[pollItemCount].events = events[socketIndex];
        items[pollItemCount].revents = 0;
        pollItemCount++;
      }

      if (pollItemCount != curEventCount) return 0;

      pollItemCount = 0;
      // waits for milliseconds, not microseconds
      int result = zmq.zmq_poll(items, curEventCount, new NativeLong(timeout));
      for (int socketIndex = 0; socketIndex < sockets.length; socketIndex++) {
        if (sockets[socketIndex] == null) continue;
        revents[socketIndex] = items[pollItemCount].revents;
        pollItemCount++;
      }
      return result;
    }

    public boolean pollin(int index) {
      return poll_mask(index, POLLIN);
    }

    public boolean pollout(int index) {
      return poll_mask(index, POLLOUT);
    }

    public boolean pollerr(int index) {
      return poll_mask(index, POLLERR);
    }

    protected Poller(Context context) {
      this(context, SIZE_DEFAULT);
    }

    protected Poller(Context context, int size) {
      this.maxEventCount = size;
      this.sockets = new Socket[maxEventCount];
      this.events = new short[maxEventCount];
      this.revents = new short[maxEventCount];
      this.freeSlots = new LinkedList<Integer>();
    }

    private boolean poll_mask(int index, int mask) {
      return !(mask <= 0 || index < 0 || index >= nextEventIndex) && (revents[index] & mask) > 0;
    }
  }
}
